File: Assignment4/EmployeeScheduleManager.java
import java.util.*;

public class EmployeeScheduleManager {
    private static final String[] DAYS = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
    private static final String[] SHIFTS = {"Morning", "Afternoon", "Evening"};
    private static final int MAX_DAYS_PER_EMPLOYEE = 5;
    private static final int MIN_EMPLOYEES_PER_SHIFT = 2;
    
    private Map<String, Employee> employees;
    private Map<String, Map<String, List<String>>> schedule; // day -> shift -> employees
    
    public EmployeeScheduleManager() {
        this.employees = new HashMap<>();
        this.schedule = new HashMap<>();
        initializeSchedule();
    }
    
    private void initializeSchedule() {
        for (String day : DAYS) {
            schedule.put(day, new HashMap<>());
            for (String shift : SHIFTS) {
                schedule.get(day).put(shift, new ArrayList<>());
            }
        }
    }
    
    public void addEmployee(String name) {
        employees.put(name, new Employee(name));
    }
    
    public void setEmployeePreference(String name, String day, String shift) {
        if (employees.containsKey(name)) {
            employees.get(name).setPreference(day, shift);
        }
    }
    
    public void generateSchedule() {
        // First pass: Fill minimum requirements (priority over preferences)
        fillMinimumRequirements();
        
        // Second pass: Assign preferred shifts (if space available)
        assignPreferredShifts();
        
        // Third pass: Resolve conflicts
        resolveConflicts();
    }
    
    private void assignPreferredShifts() {
        for (String day : DAYS) {
            for (String employeeName : employees.keySet()) {
                Employee employee = employees.get(employeeName);
                
                // Check if employee can work more days
                if (employee.getDaysWorked() >= MAX_DAYS_PER_EMPLOYEE) {
                    continue;
                }
                
                // Check if employee already has a shift this day
                if (employee.hasShiftOnDay(day)) {
                    continue;
                }
                
                String preferredShift = employee.getPreference(day);
                if (preferredShift != null && schedule.get(day).get(preferredShift).size() < 3) {
                    schedule.get(day).get(preferredShift).add(employeeName);
                    employee.assignShift(day, preferredShift);
                }
            }
        }
    }
    
    private void fillMinimumRequirements() {
        // Calculate total shifts needed: 7 days × 3 shifts × 2 employees = 42 employee-shifts
        // Available: 9 employees × 5 days = 45 employee-shifts
        // Strategy: Round-robin distribution to ensure all days get covered
        
        // Create a list of all shifts that need to be filled
        List<String[]> shiftsToFill = new ArrayList<>();
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                shiftsToFill.add(new String[]{day, shift});
            }
        }
        
        // Round-robin assignment: assign employees to shifts in rotation
        int employeeIndex = 0;
        List<String> employeeNames = new ArrayList<>(employees.keySet());
        
        for (String[] shiftInfo : shiftsToFill) {
            String day = shiftInfo[0];
            String shift = shiftInfo[1];
            List<String> shiftEmployees = schedule.get(day).get(shift);
            
            // Assign 2 employees to this shift
            for (int i = 0; i < MIN_EMPLOYEES_PER_SHIFT; i++) {
                String employeeName = findNextAvailableEmployee(day, employeeNames, employeeIndex);
                if (employeeName != null) {
                    shiftEmployees.add(employeeName);
                    employees.get(employeeName).assignShift(day, shift);
                    System.out.println("Assigned " + employeeName + " to " + day + " " + shift + " (minimum requirement)");
                    employeeIndex = (employeeIndex + 1) % employeeNames.size();
                } else {
                    System.out.println("WARNING: Cannot meet minimum requirement for " + day + " " + shift + 
                                     " (need more employees or adjust preferences)");
                    break;
                }
            }
        }
    }
    
    private String findNextAvailableEmployee(String day, List<String> employeeNames, int startIndex) {
        // Try to find an available employee starting from the given index
        for (int i = 0; i < employeeNames.size(); i++) {
            int index = (startIndex + i) % employeeNames.size();
            String employeeName = employeeNames.get(index);
            Employee employee = employees.get(employeeName);
            
            if (employee.getDaysWorked() < MAX_DAYS_PER_EMPLOYEE && !employee.hasShiftOnDay(day)) {
                return employeeName;
            }
        }
        return null;
    }
    
    private void distributeRemainingCapacity() {
        // Find employees who haven't worked 5 days yet
        List<String> availableEmployees = new ArrayList<>();
        for (String employeeName : employees.keySet()) {
            Employee employee = employees.get(employeeName);
            if (employee.getDaysWorked() < MAX_DAYS_PER_EMPLOYEE) {
                availableEmployees.add(employeeName);
            }
        }
        
        // Distribute remaining capacity across all days evenly
        for (String employeeName : availableEmployees) {
            Employee employee = employees.get(employeeName);
            int remainingDays = MAX_DAYS_PER_EMPLOYEE - employee.getDaysWorked();
            
            // Find days where this employee can work and shifts need more people
            for (String day : DAYS) {
                if (remainingDays <= 0) break;
                if (employee.hasShiftOnDay(day)) continue;
                
                for (String shift : SHIFTS) {
                    if (remainingDays <= 0) break;
                    List<String> shiftEmployees = schedule.get(day).get(shift);
                    
                    // Add employee if shift has less than 3 people (to avoid over-staffing)
                    if (shiftEmployees.size() < 3) {
                        shiftEmployees.add(employeeName);
                        employee.assignShift(day, shift);
                        System.out.println("Assigned " + employeeName + " to " + day + " " + shift + " (additional capacity)");
                        remainingDays--;
                        break; // Move to next day
                    }
                }
            }
        }
    }
    
    private String findAvailableEmployee(String day) {
        for (String employeeName : employees.keySet()) {
            Employee employee = employees.get(employeeName);
            
            // Check if employee can work more days and doesn't have shift this day
            if (employee.getDaysWorked() < MAX_DAYS_PER_EMPLOYEE && !employee.hasShiftOnDay(day)) {
                return employeeName;
            }
        }
        return null;
    }
    
    private String findEmployeeForForcedAssignment(String day) {
        // Find employee who hasn't worked 5 days yet and doesn't have shift this day
        for (String employeeName : employees.keySet()) {
            Employee employee = employees.get(employeeName);
            
            if (employee.getDaysWorked() < MAX_DAYS_PER_EMPLOYEE && !employee.hasShiftOnDay(day)) {
                return employeeName;
            }
        }
        return null;
    }
    
    private void resolveConflicts() {
        // Check for employees working more than 5 days
        for (String employeeName : employees.keySet()) {
            Employee employee = employees.get(employeeName);
            
            if (employee.getDaysWorked() > MAX_DAYS_PER_EMPLOYEE) {
                // Remove excess shifts
                List<String> shiftsToRemove = employee.getExcessShifts();
                for (String shiftInfo : shiftsToRemove) {
                    String[] parts = shiftInfo.split(":");
                    String day = parts[0];
                    String shift = parts[1];
                    
                    schedule.get(day).get(shift).remove(employeeName);
                    employee.removeShift(day, shift);
                }
            }
        }
    }
    
    public void printSchedule() {
        System.out.println("\n=== WEEKLY EMPLOYEE SCHEDULE ===");
        System.out.println("=" + "=".repeat(50));
        
        for (String day : DAYS) {
            System.out.println("\n" + day + ":");
            System.out.println("-".repeat(20));
            
            for (String shift : SHIFTS) {
                List<String> shiftEmployees = schedule.get(day).get(shift);
                System.out.printf("%-10s: %s%n", shift, 
                    shiftEmployees.isEmpty() ? "No employees assigned" : 
                    String.join(", ", shiftEmployees));
            }
        }
        
        System.out.println("\n" + "=".repeat(50));
        printEmployeeSummary();
    }
    
    private void printEmployeeSummary() {
        System.out.println("\nEMPLOYEE SUMMARY:");
        System.out.println("-".repeat(30));
        
        for (String employeeName : employees.keySet()) {
            Employee employee = employees.get(employeeName);
            System.out.printf("%-15s: %d days worked%n", employeeName, employee.getDaysWorked());
        }
    }
    
    // Employee class
    private static class Employee {
        private String name;
        private Map<String, String> preferences; // day -> preferred shift
        private Map<String, String> assignedShifts; // day -> assigned shift
        private int daysWorked;
        
        public Employee(String name) {
            this.name = name;
            this.preferences = new HashMap<>();
            this.assignedShifts = new HashMap<>();
            this.daysWorked = 0;
        }
        
        public void setPreference(String day, String shift) {
            preferences.put(day, shift);
        }
        
        public String getPreference(String day) {
            return preferences.get(day);
        }
        
        public void assignShift(String day, String shift) {
            if (!assignedShifts.containsKey(day)) {
                assignedShifts.put(day, shift);
                daysWorked++;
            }
        }
        
        public void removeShift(String day, String shift) {
            if (assignedShifts.containsKey(day) && assignedShifts.get(day).equals(shift)) {
                assignedShifts.remove(day);
                daysWorked--;
            }
        }
        
        public boolean hasShiftOnDay(String day) {
            return assignedShifts.containsKey(day);
        }
        
        public int getDaysWorked() {
            return daysWorked;
        }
        
        public List<String> getExcessShifts() {
            List<String> excess = new ArrayList<>();
            if (daysWorked > MAX_DAYS_PER_EMPLOYEE) {
                int toRemove = daysWorked - MAX_DAYS_PER_EMPLOYEE;
                List<String> allShifts = new ArrayList<>(assignedShifts.keySet());
                
                for (int i = 0; i < toRemove && i < allShifts.size(); i++) {
                    String day = allShifts.get(i);
                    String shift = assignedShifts.get(day);
                    excess.add(day + ":" + shift);
                }
            }
            return excess;
        }
    }
    
    // Main method for testing
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        EmployeeScheduleManager manager = new EmployeeScheduleManager();
        
        System.out.println("=== EMPLOYEE SCHEDULE MANAGER ===");
        System.out.println("Enter employee information and preferences:");
        System.out.println("NOTE: To meet minimum requirements (2 employees per shift per day),");
        System.out.println("you need at least 9 employees (7 days × 3 shifts × 2 employees ÷ 5 days max per employee)");
        
        // Add employees
        System.out.print("Enter number of employees (minimum 9 recommended): ");
        int numEmployees = scanner.nextInt();
        scanner.nextLine(); // consume newline
        
        for (int i = 0; i < numEmployees; i++) {
            System.out.print("Enter employee name: ");
            String name = scanner.nextLine();
            manager.addEmployee(name);
        }
        
        // Set preferences
        System.out.println("\nEnter shift preferences for each employee:");
        for (String employeeName : manager.employees.keySet()) {
            System.out.println("\nPreferences for " + employeeName + ":");
            
            for (String day : DAYS) {
                System.out.print(day + " (1=Morning, 2=Afternoon, 3=Evening, 0=None): ");
                int choice = scanner.nextInt();
                scanner.nextLine(); // consume newline
                
                if (choice >= 1 && choice <= 3) {
                    String shift = SHIFTS[choice - 1];
                    manager.setEmployeePreference(employeeName, day, shift);
                }
            }
        }
        
        // Generate and display schedule
        System.out.println("\nGenerating schedule...");
        manager.generateSchedule();
        manager.printSchedule();
        
        scanner.close();
    }
}

File: Assignment4/employee_schedule_manager.py
class EmployeeScheduleManager:
    def __init__(self):
        self.DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
        self.SHIFTS = ["Morning", "Afternoon", "Evening"]
        self.MAX_DAYS_PER_EMPLOYEE = 5
        self.MIN_EMPLOYEES_PER_SHIFT = 2
        
        self.employees = {}
        self.schedule = {}
        self._initialize_schedule()
    
    def _initialize_schedule(self):
        """Initialize the schedule data structure"""
        for day in self.DAYS:
            self.schedule[day] = {}
            for shift in self.SHIFTS:
                self.schedule[day][shift] = []
    
    def add_employee(self, name):
        """Add a new employee to the system"""
        self.employees[name] = Employee(name)
    
    def set_employee_preference(self, name, day, shift):
        """Set an employee's shift preference for a specific day"""
        if name in self.employees:
            self.employees[name].set_preference(day, shift)
    
    def generate_schedule(self):
        """Generate the weekly schedule using control structures"""
        # First pass: Fill minimum requirements (priority over preferences)
        self._fill_minimum_requirements()
        
        # Second pass: Assign preferred shifts (if space available)
        self._assign_preferred_shifts()
        
        # Third pass: Resolve conflicts
        self._resolve_conflicts()
    
    def _assign_preferred_shifts(self):
        """Assign employees to their preferred shifts"""
        for day in self.DAYS:
            for employee_name in self.employees:
                employee = self.employees[employee_name]
                
                # Check if employee can work more days (conditional)
                if employee.get_days_worked() >= self.MAX_DAYS_PER_EMPLOYEE:
                    continue
                
                # Check if employee already has a shift this day (conditional)
                if employee.has_shift_on_day(day):
                    continue
                
                preferred_shift = employee.get_preference(day)
                # Conditional check for shift availability
                if (preferred_shift is not None and 
                    len(self.schedule[day][preferred_shift]) < 3):
                    
                    self.schedule[day][preferred_shift].append(employee_name)
                    employee.assign_shift(day, preferred_shift)
    
    def _fill_minimum_requirements(self):
        """Ensure minimum employee requirements are met for each shift"""
        # Calculate total shifts needed: 7 days × 3 shifts × 2 employees = 42 employee-shifts
        # Available: 9 employees × 5 days = 45 employee-shifts
        # Strategy: Round-robin distribution to ensure all days get covered
        
        # Create a list of all shifts that need to be filled
        shifts_to_fill = []
        for day in self.DAYS:
            for shift in self.SHIFTS:
                shifts_to_fill.append((day, shift))
        
        # Round-robin assignment: assign employees to shifts in rotation
        employee_index = 0
        employee_names = list(self.employees.keys())
        
        for day, shift in shifts_to_fill:
            shift_employees = self.schedule[day][shift]
            
            # Assign 2 employees to this shift
            for i in range(self.MIN_EMPLOYEES_PER_SHIFT):
                employee_name = self._find_next_available_employee(day, employee_names, employee_index)
                if employee_name is not None:
                    shift_employees.append(employee_name)
                    self.employees[employee_name].assign_shift(day, shift)
                    print(f"Assigned {employee_name} to {day} {shift} (minimum requirement)")
                    employee_index = (employee_index + 1) % len(employee_names)
                else:
                    print(f"WARNING: Cannot meet minimum requirement for {day} {shift} (need more employees or adjust preferences)")
                    break
    
    def _distribute_remaining_capacity(self):
        """Distribute remaining employee capacity evenly across all days"""
        # Find employees who haven't worked 5 days yet
        available_employees = []
        for employee_name in self.employees:
            employee = self.employees[employee_name]
            if employee.get_days_worked() < self.MAX_DAYS_PER_EMPLOYEE:
                available_employees.append(employee_name)
        
        # Distribute remaining capacity across all days evenly
        for employee_name in available_employees:
            employee = self.employees[employee_name]
            remaining_days = self.MAX_DAYS_PER_EMPLOYEE - employee.get_days_worked()
            
            # Find days where this employee can work and shifts need more people
            for day in self.DAYS:
                if remaining_days <= 0:
                    break
                if employee.has_shift_on_day(day):
                    continue
                
                for shift in self.SHIFTS:
                    if remaining_days <= 0:
                        break
                    shift_employees = self.schedule[day][shift]
                    
                    # Add employee if shift has less than 3 people (to avoid over-staffing)
                    if len(shift_employees) < 3:
                        shift_employees.append(employee_name)
                        employee.assign_shift(day, shift)
                        print(f"Assigned {employee_name} to {day} {shift} (additional capacity)")
                        remaining_days -= 1
                        break  # Move to next day
    
    def _find_available_employee(self, day):
        """Find an employee available for a specific day"""
        for employee_name in self.employees:
            employee = self.employees[employee_name]
            
            # Conditional checks for availability
            if (employee.get_days_worked() < self.MAX_DAYS_PER_EMPLOYEE and 
                not employee.has_shift_on_day(day)):
                return employee_name
        return None
    
    def _find_next_available_employee(self, day, employee_names, start_index):
        """Find an available employee starting from the given index"""
        # Try to find an available employee starting from the given index
        for i in range(len(employee_names)):
            index = (start_index + i) % len(employee_names)
            employee_name = employee_names[index]
            employee = self.employees[employee_name]
            
            if (employee.get_days_worked() < self.MAX_DAYS_PER_EMPLOYEE and 
                not employee.has_shift_on_day(day)):
                return employee_name
        return None
    
    def _find_employee_for_forced_assignment(self, day):
        """Find employee for forced assignment to meet minimum requirements"""
        for employee_name in self.employees:
            employee = self.employees[employee_name]
            
            # Find employee who hasn't worked 5 days yet and doesn't have shift this day
            if (employee.get_days_worked() < self.MAX_DAYS_PER_EMPLOYEE and 
                not employee.has_shift_on_day(day)):
                return employee_name
        return None
    
    def _resolve_conflicts(self):
        """Resolve scheduling conflicts"""
        for employee_name in self.employees:
            employee = self.employees[employee_name]
            
            # Conditional check for over-scheduling
            if employee.get_days_worked() > self.MAX_DAYS_PER_EMPLOYEE:
                # Remove excess shifts
                shifts_to_remove = employee.get_excess_shifts()
                for shift_info in shifts_to_remove:
                    day, shift = shift_info.split(":")
                    self.schedule[day][shift].remove(employee_name)
                    employee.remove_shift(day, shift)
    
    def print_schedule(self):
        """Print the weekly schedule in a readable format"""
        print("\n=== WEEKLY EMPLOYEE SCHEDULE ===")
        print("=" * 50)
        
        # Loop through days
        for day in self.DAYS:
            print(f"\n{day}:")
            print("-" * 20)
            
            # Loop through shifts
            for shift in self.SHIFTS:
                shift_employees = self.schedule[day][shift]
                if not shift_employees:
                    print(f"{shift:<10}: No employees assigned")
                else:
                    print(f"{shift:<10}: {', '.join(shift_employees)}")
        
        print("\n" + "=" * 50)
        self._print_employee_summary()
    
    def _print_employee_summary(self):
        """Print summary of employee work days"""
        print("\nEMPLOYEE SUMMARY:")
        print("-" * 30)
        
        for employee_name in self.employees:
            employee = self.employees[employee_name]
            print(f"{employee_name:<15}: {employee.get_days_worked()} days worked")


class Employee:
    def __init__(self, name):
        self.name = name
        self.preferences = {}  # day -> preferred shift
        self.assigned_shifts = {}  # day -> assigned shift
        self.days_worked = 0
    
    def set_preference(self, day, shift):
        """Set shift preference for a specific day"""
        self.preferences[day] = shift
    
    def get_preference(self, day):
        """Get shift preference for a specific day"""
        return self.preferences.get(day)
    
    def assign_shift(self, day, shift):
        """Assign a shift to the employee"""
        if day not in self.assigned_shifts:
            self.assigned_shifts[day] = shift
            self.days_worked += 1
    
    def remove_shift(self, day, shift):
        """Remove a shift assignment"""
        if day in self.assigned_shifts and self.assigned_shifts[day] == shift:
            del self.assigned_shifts[day]
            self.days_worked -= 1
    
    def has_shift_on_day(self, day):
        """Check if employee has a shift on a specific day"""
        return day in self.assigned_shifts
    
    def get_days_worked(self):
        """Get total number of days worked"""
        return self.days_worked
    
    def get_excess_shifts(self):
        """Get list of excess shifts to remove"""
        excess = []
        if self.days_worked > 5:  # MAX_DAYS_PER_EMPLOYEE
            to_remove = self.days_worked - 5
            all_shifts = list(self.assigned_shifts.keys())
            
            # Loop to collect excess shifts
            for i in range(min(to_remove, len(all_shifts))):
                day = all_shifts[i]
                shift = self.assigned_shifts[day]
                excess.append(f"{day}:{shift}")
        
        return excess


def main():
    """Main function to run the employee schedule manager"""
    manager = EmployeeScheduleManager()
    
    print("=== EMPLOYEE SCHEDULE MANAGER ===")
    print("Enter employee information and preferences:")
    
    # Add employees
    num_employees = int(input("Enter number of employees: "))
    
    for i in range(num_employees):
        name = input("Enter employee name: ")
        manager.add_employee(name)
    
    # Set preferences
    print("\nEnter shift preferences for each employee:")
    for employee_name in manager.employees:
        print(f"\nPreferences for {employee_name}:")
        
        for day in manager.DAYS:
            print(f"{day} (1=Morning, 2=Afternoon, 3=Evening, 0=None): ", end="")
            choice = int(input())
            
            if 1 <= choice <= 3:
                shift = manager.SHIFTS[choice - 1]
                manager.set_employee_preference(employee_name, day, shift)
    
    # Generate and display schedule
    print("\nGenerating schedule...")
    manager.generate_schedule()
    manager.print_schedule()


if __name__ == "__main__":
    main()
